experimental/puppeteer-firefox/lib/DOMWorld.js

```diff
@@ -409,24 +409,52 @@ class DOMWorld {
      * @return {?Node|boolean}
      */
     function predicate(selectorOrXPath, isXPath, waitForVisible, waitForHidden) {
-      const node = isXPath
-        ? document.evaluate(selectorOrXPath, document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue
-        : document.querySelector(selectorOrXPath);
-      if (!node)
-        return waitForHidden;
-      if (!waitForVisible && !waitForHidden)
-        return node;
-      const element = /** @type {Element} */ (node.nodeType === Node.TEXT_NODE ? node.parentElement : node);
-
-      const style = window.getComputedStyle(element);
-      const isVisible = style && style.visibility !== 'hidden' && hasVisibleBoundingBox();
```

Comment:
@aslushnikov you're very quick to close this out... but if you look at the code here, you'll realize that it originally already uses the so called `magic` to determine visibility (as you can see by `const isVisible = style && style.visibility !== 'hidden' && hasVisibleBoundingBox();`).

My fix merely just loops over all elements on page instead of stopping at the **first** element to determine `visibility`...

could you take a closer look at this before closing it out. thx!

Reply:
> already uses the so called magic to determine visibility (as you can see by const isVisible = style && style.visibility !== 'hidden' && hasVisibleBoundingBox();).

@razorman8669  Sure, but this magic is scoped to determining visibility and doesn't affect element selection.

Everywhere throughout our API we rely on CSS Selectors to exclusively determine the element we act upon. We never try to be clever and "find another element" if something goes wrong with the one user gives us. And we shouldn't: users can always provide a more specific css selector if they need to.

The approach when we use a combination of a css selector and options to determine the target element doesn't play nice with the rest of our API and our conventions.

Reply:
aha, I see what you're saying now and that makes sense.  more specific CSS selector it is!  thanks for clarifying your feedback.